### PR TITLE
DEV: Adds checks for 'prioritize_full_name_in_ux'

### PR DESCRIPTION
**Pull request by brrusselburg**
Equivalent core PR [here](https://github.com/discourse/discourse/pull/30583).

Ensures full names get passed through if the hidden `prioritize_full_name_in_ux` setting is `true`.

Before (with setting on):
Screenshot 2025-02-12 at 12 04 39 PM
Screenshot 2025-02-12 at 12 07 02 PM

After (with setting on):
Screenshot 2025-02-12 at 12 02 43 PM
Screenshot 2025-02-12 at 12 06 16 PM

### CODE REVIEW

**Review by martin-brennan — COMMENTED** [review]
This does seem on the right track, did you want me to dig in to any of the failing stuff, or are you on top of it?

**Comment by brrusselburg** [comment]
> This does seem on the right track, did you want me to dig in to any of the failing stuff, or are you on top of it?

Sweet, thanks! I poked around more today and found a couple things I forgot to update (mainly the yaml) but still ran into the same block. The reaction consolidation is happening in the ruby spec, so I assume the consolidation is getting lost in the js somewhere.

Screenshot 2025-02-10 at 5 11 00 PM

File: spec/system/reaction_notifications_spec.rb [hunk 1]
```diff
@@ -63,4 +63,36 @@
...
+
+    it "renders reaction notifications with full names" do
+      visit("/")
+      user_menu.open
+
+      labels = page.all("#quick-access-all-notifications .notification.reaction .item-label")
+
+      expect(labels[0]).to have_text(
+        I18n.t(
+          "js.notifications.reaction_multiple_users.one",
```
Comment (brrusselburg) [inline spec/system/reaction_notifications_spec.rb:82]: This fails saying it doesn't exist, but.... (see other comment)

File: config/locales/client.en.yml [hunk 2]
```diff
@@ -28,12 +28,5 @@ en:
       reaction_multiple_users:
```
Comment (brrusselburg) [inline config/locales/client.en.yml:28]: ...this exists? I need a second pair of eyes

File: config/locales/client.en.yml [hunk 3]
```diff
@@ -28,12 +28,5 @@ en:
...
         one: "%{username} and %{count} other"
```
Comment (brrusselburg) [inline config/locales/client.en.yml:29]: cc @jjaffeux -- I had the wrong line

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs [hunk 4]
```diff
@@ -128,21 +128,25 @@ function initializeDiscourseReactions(api) {
 
         get label() {
           const count = this.notification.data.count;
-          const username = this.username;
+          const nameOrUsername = this.siteSettings.prioritize_full_name_in_ux
+            ? this.notification.data.display_name
+            : this.notification.data.display_username;
```
Comment (OsamaSayegh) [inline assets/javascripts/discourse/initializers/discourse-reactions.gjs]:
```suggestion
            : this.username;
```

It currently uses `this.username` so we should keep it. They're practically the same, the only difference is that `this.username` has the `formatUsername` function applied to it whereas `this.notification.data.display_username` is the raw username.

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js [hunk 5]
```diff
@@ -241,3 +241,130 @@ acceptance("Discourse Reactions - Notifications", function (needs) {
     );
   });
 });
+
+acceptance(
+  "Discourse Reactions - Notifications | Full Name Setting On",
+  function (needs) {
+    needs.user({ redesigned_user_menu_enabled: true });
```
Comment (OsamaSayegh) [inline test/javascripts/acceptance/discourse-reactions-notifications-test.js]: This shouldn't be needed anymore. This was a feature flag to enable the notifications menu when it was still experimental, but that's not the case anymore and the feature flag is long gone.

File: spec/system/reaction_notifications_spec.rb [hunk 6]
```diff
@@ -63,4 +63,33 @@
     )
     expect(labels[2]).to have_text(acting_user_1.username)
   end
+
+  context "when prioritize full names in ux site setting is on" do
+    before do
+      SiteSetting.prioritize_full_name_in_ux = true
+      SiteSetting.discourse_reactions_enabled = true
```
Comment (OsamaSayegh) [inline spec/system/reaction_notifications_spec.rb]: I'd move this into a top level `before` block since every test here needs the plugin to be enabled.

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js [hunk 7]
```diff
@@ -241,3 +241,130 @@ acceptance("Discourse Reactions - Notifications", function (needs) {
...
+      assert.strictEqual(
+        notifications[1].textContent.replaceAll(/\s+/g, " ").trim(),
+        `${i18n("notifications.reaction_2_users", {
+          username: "Bruce Wayne I",
+          username2: "Brucer Wayner II",
+        })} Topic with 2 likes (total) from 2 users`,
+        "notification for 2 likes from 2 users has the right content"
+      );
+
+      // this one isn't receiving username value for the translation for some reason - solution should also fix the rails test error
```
Comment (OsamaSayegh) [inline test/javascripts/acceptance/discourse-reactions-notifications-test.js]: Leftover comment?

**Review by OsamaSayegh — COMMENTED** [review]
Looks pretty good, just a few minor comments and it should be ready to go 🚀

File: spec/system/reaction_notifications_spec.rb (hunk 6 above)
Comment (brrusselburg) [reply spec/system/reaction_notifications_spec.rb]: I removed it, there's one at the top already 🧹

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js (hunk 5 above)
Comment (brrusselburg) [reply test/javascripts/acceptance/discourse-reactions-notifications-test.js]: Removed it but then the spec failed, so I put it back in

File: app/services/discourse_reactions/reaction_notification.rb [hunk 8]
```diff
@@ -18,7 +18,7 @@ def create
         return
       end
 
-      opts = { user_id: @user.id, display_username: @user.username }
+      opts = { user_id: @user.id, display_username: @user.username, display_name: @user.name }
```
Comment (cvx) [inline app/services/discourse_reactions/reaction_notification.rb:21]: a general question for the entire PR: does it fallback to the username if `name` is empty?

Comment (brrusselburg) [reply app/services/discourse_reactions/reaction_notification.rb:21]: Just tested and it does _not_ 😭 thank you @CvX!

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js (hunk 5 above)
Comment (OsamaSayegh) [reply test/javascripts/acceptance/discourse-reactions-notifications-test.js]: Sorry I wasn't clear here — the `redesigned_user_menu_enabled` argument shouldn't be needed, but the `needs.user()` call needs to stay since it's what sets up the current user object for the test. Without `needs.user()`, we get no current user for the test and it fails because there's no notifications menu for anonymous users.

Comment (brrusselburg) [reply test/javascripts/acceptance/discourse-reactions-notifications-test.js]: OHH, I see! Thank you very much for clarifying :)

File: assets/javascripts/discourse/widgets/discourse-reactions-list-emoji.js [hunk 9]
```diff
@@ -41,7 +41,12 @@ export default createWidget("discourse-reactions-list-emoji", {
     } else {
       users.slice(0, DISPLAY_MAX_USERS).forEach((user) => {
         let displayName;
-        if (this.siteSettings.prioritize_username_in_ux) {
+        if (
+          !this.siteSettings.prioritize_username_in_ux &&
+          this.siteSettings.prioritize_full_name_in_ux
+        ) {
+          displayName = user.name;
+        } else if (this.siteSettings.prioritize_username_in_ux) {
```
Comment (cvx) [inline assets/javascripts/discourse/widgets/discourse-reactions-list-emoji.js]: this doesn't fallback to username?

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs [hunk 10]
```diff
@@ -128,21 +128,26 @@ function initializeDiscourseReactions(api) {
 
         get label() {
           const count = this.notification.data.count;
-          const username = this.username;
+          const nameOrUsername = this.siteSettings.prioritize_full_name_in_ux
+            ? this.notification.data.display_name ||
+              this.notification.data.username
```
Comment (OsamaSayegh) [inline assets/javascripts/discourse/initializers/discourse-reactions.gjs]:
```suggestion
              this.username
```

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs [hunk 11]
```diff
@@ -128,21 +128,26 @@ function initializeDiscourseReactions(api) {
...
+            : this.username;
```
Comment (cvx) [inline assets/javascripts/discourse/initializers/discourse-reactions.gjs]: what's the diff between `this.notification.data.username` and `this.username`? and should both be used there? (depending on the site setting? 🤔)

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs [hunk 12]
```diff
@@ -128,21 +128,26 @@ function initializeDiscourseReactions(api) {
...
-
           if (count > 2) {
             return i18n("notifications.reaction_multiple_users", {
-              username,
+              username: nameOrUsername,
               count: count - 1,
             });
           } else {
+            const nameOrUsername2 = this.siteSettings.prioritize_full_name_in_ux
+              ? this.notification.data.name2 || this.notification.data.username2
```
Comment (OsamaSayegh) [inline assets/javascripts/discourse/initializers/discourse-reactions.gjs]:
```suggestion
              ? this.notification.data.name2 || formatUsername(this.notification.data.username2)
```

(or possibly extract `formatUsername(this.notification.data.username2)` into a variable and use it in the 2 spots in this ternary condition, up to you)

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js [hunk 13]
```diff
@@ -241,3 +241,129 @@ acceptance("Discourse Reactions - Notifications", function (needs) {
...
+                name2: "Roronoa Zoro",
+                count: 4,
+              },
+            },
+          ],
+        });
+      });
+    });
+
+    test("reaction notifications with full name site setting on", async (assert) => {
```
Comment (cvx) [inline test/javascripts/acceptance/discourse-reactions-notifications-test.js]: it's better to stick to the non-arrow functions as test methods (doesn't matter in this case, but if you were to use `this`, it wouldn't work)

```suggestion
    test("reaction notifications with full name site setting on", async function (assert) {
```

File: spec/system/reaction_notifications_spec.rb [hunk 14]
```diff
@@ -63,4 +63,29 @@
     )
     expect(labels[2]).to have_text(acting_user_1.username)
   end
+
+  context "when prioritize full names in ux site setting is on" do
+    before { SiteSetting.prioritize_full_name_in_ux = true }
+
+    it "renders reaction notifications with full names" do
```
Comment (OsamaSayegh) [inline spec/system/reaction_notifications_spec.rb:81]: Now that we've added the username fallback, I think it's worth adding a test case for when the full name setting is on but users have a null name to test that we actually fallback to the username.

**Review by OsamaSayegh — COMMENTED** [review]
we're this 🤏 close to merging the PR, just a couple more things 🚀

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js [hunk 15]
```diff
@@ -241,3 +241,129 @@ acceptance("Discourse Reactions - Notifications", function (needs) {
...
+                original_post_type: 1,
+                original_username: "krus",
+                revision_number: null,
+                display_username: "krus",
+                display_name: "Brucer Wayner II",
+                reaction_icon: "heart",
+                previous_notification_id: 933,
+                count: 1,
+              },
+            },
```
Comment (cvx) [inline test/javascripts/acceptance/discourse-reactions-notifications-test.js]: if this notification isn't referenced in the test we can remove it imo

File: test/javascripts/acceptance/discourse-reactions-notifications-test.js [hunk 16]
```diff
@@ -241,3 +241,129 @@ acceptance("Discourse Reactions - Notifications", function (needs) {
...
+      );
+
+      assert.strictEqual(
+        notifications[1].textContent.replaceAll(/\s+/g, " ").trim(),
+        `${i18n("notifications.reaction_2_users", {
+          username: "Bruce Wayne I",
+          username2: "Brucer Wayner II",
+        })} Topic with 2 likes (total) from 2 users`,
+        "notification for 2 likes from 2 users has the right content"
+      );
```
Comment (cvx) [inline test/javascripts/acceptance/discourse-reactions-notifications-test.js]: both assertions could be rewritten like this, avoiding the need for `queryAll` (I want to deprecate it at some point, it's mostly gone from core), manual reaching into `textContent`, regex replace, etc.

```suggestion
      assert
        .dom("li.notification.reaction:nth(1) a")
        .hasText(
          /Bruce Wayne I, Brucer Wayner II\s+Topic with 2 likes (total) from 2 users/,
        );
```

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs (hunk 11 above)
Comment (OsamaSayegh) [reply assets/javascripts/discourse/initializers/discourse-reactions.gjs]: `this.username` is a shorthand for `formatUsername(this.notification.data.username)` defined in `NotificationTypeBase`.

Comment (brrusselburg) [reply assets/javascripts/discourse/initializers/discourse-reactions.gjs]: turns out this was an error on my part, because on the notif object it's referred to as `display_username`. There isn't a `.username` on the notification object.

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs (hunk 10 above)
Comment (brrusselburg) [reply assets/javascripts/discourse/initializers/discourse-reactions.gjs]: Any reason this shouldn't be `this.notification.data.display_username`? Turns out `.username` doesn't exist on the notification object, so the way I had it wouldn't have worked anyway.

Comment (OsamaSayegh) [reply assets/javascripts/discourse/initializers/discourse-reactions.gjs]: Yes `this.notification.data.username` doesn't exist, however `this.username` is just a [getter](https://github.com/discourse/discourse/blob/f744852f105f2ac355af843b646fb2362cacc46d/app/assets/javascripts/discourse/app/lib/notification-types/base.js#L113-L115) that returns `this.notification.data.display_username` with `formatUsernaem()` applied on it. Is `this.username` not returning anything in your tests?

File: assets/javascripts/discourse/widgets/discourse-reactions-list-emoji.js (hunk 9 above)
Comment (brrusselburg) [reply assets/javascripts/discourse/widgets/discourse-reactions-list-emoji.js]: oop, missed this one. thanks!

File: assets/javascripts/discourse/initializers/discourse-reactions.gjs (hunk 10 above)
Comment (brrusselburg) [reply assets/javascripts/discourse/initializers/discourse-reactions.gjs]: Ahh gotcha thank you! Nah it worked :)

**Review by blake-discourse — APPROVED** [review]
